File: examples/java-threads/Example8_Solution.java
// =================================================================
//
// File: Example8_Solution.java
// Author: Pedro Perez
// Description: This file implements the enumeration sort algorithm.
// 				The time this implementation takes will be used as the
//				basis to calculate the improvement obtained with
//				parallel technologies.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

import java.util.Arrays;

public class Example8_Solution extends Thread {
	private static final int SIZE = 100_000;
	private int array[], temp[], start, end;

	public Example8_Solution(int src[], int dst[], int start, int end) {
		this.array = src;
		this.temp = dst;
		this.start = start;
		this.end = end;
	}

	public void run() {
		int count;

		for (int i = start; i < end; i++) {
			count = 0;
			for (int j = 0; j < array.length; j++) {
				if (array[j] < array[i]) {
					count++;
				} else if (array[i] == array[j] && j < i) {
					count++;
				}
			}
			temp[count] = array[i];
		}
	}

	public static void main(String args[]) {
		int array[] = new int[SIZE];
		int temp[] = new int[SIZE];
		long startTime, stopTime;
		double ms;
		Example8_Solution threads[];
		int block;

		Utils.randomArray(array);
		Utils.displayArray("before", array);

		block = array.length / Utils.MAXTHREADS;
		threads = new Example8_Solution[Utils.MAXTHREADS];

		System.out.printf("Starting with %d threads...\n", Utils.MAXTHREADS);
		ms = 0;
		for (int j = 1; j <= Utils.N; j++) {
			for (int i = 0; i < threads.length; i++) {
				if (i != threads.length - 1) {
					threads[i] = new Example8_Solution(array, temp, (i * block), ((i + 1) * block));
				} else {
					threads[i] = new Example8_Solution(array, temp, (i * block), array.length);
				}
			}

			startTime = System.currentTimeMillis();
			for (int i = 0; i < threads.length; i++) {
				threads[i].start();
			}
			for (int i = 0; i < threads.length; i++) {
				try {
					threads[i].join();
				} catch (InterruptedException e) {
					e.printStackTrace();
				}
			}
			stopTime = System.currentTimeMillis();
			ms +=  (stopTime - startTime);
		}
		Utils.displayArray("after", temp);
		System.out.printf("avg time = %.5f\n ms", (ms / Utils.N));
	}
}

File: examples/06-openmp/example8_solution.c
// =================================================================
//
// File: example7.c
// Author(s):
// Description: This file contains the code that implements the
//				enumeration sort algorithm using OpenMP.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "utils.h"

#define SIZE 10000

void enumeration_sort(int *array, int size, int copy) {
	int *temp = (int *) malloc(sizeof(int) * size);
	int i, j, count;

  #pragma omp parallel for shared(array, temp, size) private(j, count)
	for (i = 0; i < size; i++) {
		count = 0;
		for (j = 0; j < size; j++) {
			if (array[j] < array[i]) {
				count++;
			} else if (array[i] == array[j] && j < i) {
				count++;
			}
		}
		temp[count] = array[i];
	}
	if (copy) {
		memcpy(array, temp, sizeof(int) * size);
	}
	free(temp);
}

int main(int argc, char* argv[]) {
	int i, *a;
	double ms;

	a = (int*) malloc(sizeof(int) * SIZE);
	random_array(a, SIZE);
	display_array("before", a);

	printf("Starting...\n");
	ms = 0;
	for (i = 0; i < N; i++) {
		start_timer();

		enumeration_sort(a, SIZE, (i == (N - 1)));

		ms += stop_timer();
	}
	display_array("after", a);
	printf("avg time = %.5lf ms\n", (ms / N));

	free(a);
	return 0;
}
